docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,205 @@
+  If it finds that any of the external endpoints are not healthy while the internal endpoints are, then it [scales down](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/probe-configmap.yaml#L22) the corresponding shoot's `kube-controller-manager` to `0`.
+  This is to prevent the `NodeLifecycle` controller of the `kube-controller-manager` from [marking](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L860) all the `nodes` and their `pods` in the cluster as `NotReady` because the cause of the `nodes` being not ready has probably nothing to do with the `nodes` themselves.
+  It is probably because the `kube-apiserver` endpoint is not reachable externally due to loadbalancer or some other network issue.
+  - The `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the `dependency-watchdog` watches the Kubernetes `endpoints` of `etcd` and `kube-apiserver` of all the shoots (hosted in the seed cluster).
+  If any of the endpoints transition from unhealthy to healthy it restarts (deletes) the corresponding [downstream](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/endpoint-configmap.yaml#L41) [dependent](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/endpoint-configmap.yaml#L16) `pods` but only if they are in `CrashLoopBackoff`.
+  This is to speed up the cluster recovery from `etcd` or `kube-apiserver` unavailability.
+  - The `etcd` `readinessProbe` [points](https://github.com/gardener/gardener/blob/master/charts/seed-controlplane/charts/etcd/templates/etcd-statefulset.yaml#L43) to the health of the [etcd-backup-restore](https://github.com/gardener/etcd-backup-restore) [sidecar](https://github.com/gardener/gardener-extensions/blob/master/pkg/webhook/controlplane/etcd.go#L44).
+  This makes the `etcd` of the control-plane of the shoot clusters unavailable if the backup infrastructure is unavailable for any reason (even if the `etcd` container itself is healthy).
+  - The `machine-controller-manager` can in principle (it currently doesn't) notice issues in the cloud-provider that other Kubernetes and or Gardener controllers do not notice such as machines not joining in the given timeout period, cloud-provider API requests failing etc.
+
+The distributed nature of such information and action is in itself not a *bad thing*.
+But if the information is implicit and isolated within different controllers then there is a hard coupling between the sources of the information and the corrective actions that are taken which **is** a *bad thing*.
+
+## Goals
+
+- As a distributed system, the components that monitor health conditions that have cluster-wide implication should be decoupled from the components that take remedial action.
```

Comment:
> this sound to me like anti kubernetes pattern.

How exactly?

> A monitoring stack(Prometheus+Alert Manager+Grafana) is already available for every shoot and seed cluster. What do you want to add/enhance here and how it will affected the existing solution?

I am not suggesting the replacement of Prometheus + Alert Manager + Grafana. They are working as they are designed and serving their purpose well. The alerts part of the possible actions is only a minor goal. If some conditions or alerts require more domain logic than Prometheus or Alert Manager can support, then the alerts based on the conditions could be useful. Otherwise, Prometheus + Alert Manager it is.